cpp/src/arrow/compute/kernels/scalar_string_ascii.cc

```diff
@@ -2347,6 +2357,126 @@ void AddAsciiStringExtractRegex(FunctionRegistry* registry) {
   }
   DCHECK_OK(registry->AddFunction(std::move(func)));
 }
+class ExtractRegexSpanData : public ExtractRegexData {
```

Comment:
Hmm, instead of the following class diagram:
```
ExtractRegexData (defines ResolveOutputType)
|---- ExtractRegexSpanData (redefines ResolveOutputType)
```

I think it would be more consistent to have:
```
BaseExtractRegexData
|---- ExtractRegexData (defines ResolveOutputType)
|---- ExtractRegexSpanData (redefines ResolveOutputType)
```